`more`: reduce memory usage a bit

tertsdiepraam opened the pull request:
Work towards (but definitely does not close): https://github.com/uutils/coreutils/issues/6397

This just reduces the number of strings that are allocated. We're still at 5x the file though. Even though technically, we don't need to keep the file in memory at all.

BenWiederhake reviewed (approved):
LGTM! Can you do something like `src/uu/shuf/BENCHMARKING.md` and write down how exactly you measure our progress?

For this PR it's obvious that this should reduce memory consumption, so it should be trivial to prove it. For future PRs, I'm afraid that an "optimization" worsens our memory consumption and we don't notice it.

tertsdiepraam commented:
Changes since last push:
- Changed some u16 to usize to reduce conversions
- Made the pattern search take &str instead of Option
Nothing to improve performance, but cleans up a bit.

sylvestre commented:
Clippy is unhappy:
```

error: useless conversion to the same type: `usize`
   --> src/uu/more/src/more.rs:472:29
    |
472 |             .saturating_add(self.content_rows.into())
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.content_rows`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion
    = note: `-D clippy::useless-conversion` implied by `-D warnings`
    = help: to override `-D warnings` add `#[allow(clippy::useless_conversion)]`

error: useless conversion to the same type: `usize`
   --> src/uu/more/src/more.rs:488:58
    |
488 |         self.upper_mark = self.upper_mark.saturating_add(self.content_rows.into());
    |                                                          ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.content_rows`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

error: useless conversion to the same type: `usize`
   --> src/uu/more/src/more.rs:527:49
    |
527 |             .min(self.upper_mark.saturating_add(self.content_rows.into()));
    |                                                 ^^^^^^^^^^^^^^^^^^^^^^^^ help: consider removing `.into()`: `self.content_rows`
    |
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#useless_conversion

```

sylvestre commented:
@tertsdiepraam ping ? :)
